numpy/core/src/umath/loops_fast.dispatch.c.src

```diff
@@ -0,0 +1,177 @@
+/* -*- c -*- */
+/*
+ * vim:syntax=c
+ */
+/*@targets
+ ** $autovec baseline
+ ** sse2 avx2 avx512_skx
```

Comment:
@mattip, @rgommers, Adding `avx512_skx` causes increase of the binary size, should I remove it and keeping only `avx2`,  `sse2`.

Reply:
@seiko2plus it's difficult to find out what `skx` even is. From https://www.fool.com/investing/general/2014/12/10/intel-corporation-is-doing-something-strange-with.aspx I think it's "Skylake Xeon" which may or may not be different from Skylake. Either way, that's a hardware architecture thing. For AVX-512 this is the set of flavors I had in mind:

![image](https://user-images.githubusercontent.com/98330/77857532-5d598080-71fe-11ea-934f-501ca582697e.png)

From that I concluded that the `F` and/or `CD` subsets are the only relevant ones.

> Adding `avx512_skx` causes an increase of the binary size,

depends on the binary size vs performance trade-off I'd think. how much does the binary size increase?

Reply:
> depends on the binary size vs performance trade-off I'd think. how much does the binary size increase?

It wasn't easy for me to determine the performance changes through **ASV**, I had to create a standalone script #15987.

Well, I think we can temporarily keep `AVX512_SKX` and `AVX2` until replacing them later with universal intrinsics, since usually the average of end-users are counting on compiled packages, and most of them not gonna have a custom build that changes the `--cpu-baseline` to `native` or `AVX512_SKX`.

First, auto-vectorized generated machine code `AVX512_SKX` and `AVX2` are badly handle different sizes of strides, check benchmark [baseline[SSE3] vs (AVX2 and AVX512_SKX) (strides 2)](https://github.com/seiko2plus/just4share/blob/master/numpy/pr13516_bench/results/strides_2/BASELINE.md).

But on the other side, if the user didn't change the stride size, the performance gain almost good enough to force us to keep them until we replace them with universal intrinsics.
check benchmark [baseline[SSE3] vs (AVX2 and AVX512_SKX) (strides 1)](https://github.com/seiko2plus/just4share/blob/master/numpy/pr13516_bench/results/strides_1/BASELINE.md).

**Back to your question ` how much does the binary size increase?`**

Having `AVX512_SKX` as a dispatch-able feature will increase the binary size of `_multiarray_umath.cpython-37m-x86_64-linux-gnu.so` by **440kb** that if the debugging symbols are stripped otherwise the increase will easily hit the **5mb**, that may be the main goal behind #16110.

**And `performance trade-off `**?

The changes are good enough to keep it, the  average of ratios around (1.20:1.6) compared to AVX2 but only if stride size is not changed, otherwise no performance gain also slight to semi-heavy lose of performance in a few ufuncs, check the following benchmark tables for more clarification: 

- [AVX2 vs AVX512_SKX (strides 1)](https://github.com/seiko2plus/just4share/blob/master/numpy/pr13516_bench/results/strides_1/AVX2.md) 
- [AVX2 vs AVX512_SKX (strides 2)](https://github.com/seiko2plus/just4share/blob/master/numpy/pr13516_bench/results/strides_2/AVX2.md) 
- [Full report](https://github.com/seiko2plus/just4share/blob/master/numpy/pr13516_bench/)

>  it's difficult to find out what `skx` even is. From https://www.fool.com/investing/general/2014/12/10/intel-corporation-is-doing-something-strange-with.aspx 

By `AVX512_SKX` we mean a group of AVX512 features that comes with Skylake **Server**  and **X** stand for **xeon** chips.  **Note**: all features groups have extra underscore to separate them from the single features.

**Skylake server**  chips are different than other PC chips that have the same architecture code name. the main difference is the **Skylake X** comes with AVX512 with the full support of integer operations **(AVX512/F,CD,BW,DQ,VL)** while the PC version of **Skylake**  only supports **AVX2/FMA3**.

Reply:
Thanks for the detailed feedback @seiko2plus! I agree with your assessment, ~20% gain in many functions for ~440kb binary size increase is definitely worth keeping. Stride 1 should be much common that stride 2, the few functions that lose performance for stride 2 are not a blocker I'd think.

Reply:
@rgommers, I will consider it resolved, thank u